File: java/algorithm/divideandconquer/PowxN.java
package algorithm.divideandconquer;

/**
 * @author roseduan
 * @time 2020/9/26 9:26 上午
 * @description 求x的n次方
 */
public class PowxN {

    public double myPow(double x, int n) {
        return n < 0 ? 1 / helper(x, -n) : helper(x, n);
    }

    private double helper(double x, long n) {
        if (n == 0) {
            return 1;
        }
        double res = helper(x, n / 2);
        return n % 2 == 0 ? res * res : x * res * res;
    }

}

File: python/divideconquer/powx_n.py

# x的n次方

class Solution:

    # 分治递归
    def myPow_1(self, x: float, n: int) -> float:
        return self.helper(x, n) if n > 0 else 1 / self.helper(x, -n)

    def helper(self, x, n) -> float:
        if n == 0:
            return 1
        if n == 1:
            return x
        res = self.helper(x, n // 2)
        return res * res if n % 2 == 0 else res * res * x
